File: internal/images/contexts/_shared/models/hub.py
"""Pydantic models for hub configuration and provenance tracking."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator


class HubSource(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    type: Literal["git"] = "git"
    url: str
    branch: str = "main"


class HubConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sources: list[HubSource] = []

    @field_validator("sources", mode="before")
    @classmethod
    def coerce_none_to_list(cls, v: object) -> object:
        return v if v is not None else []


class AgencyConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    hub: HubConfig = HubConfig()


class HubInstalledEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")

    component: str
    kind: str
    source: str
    commit_sha: str
    installed_at: str

File: internal/models/hub.go
// agency-gateway/internal/models/hub.go
package models

// HubSource represents a git-based hub source.
type HubSource struct {
	Name   string `yaml:"name" validate:"required"`
	Type   string `yaml:"type" validate:"required,oneof=git" default:"git"`
	URL    string `yaml:"url" validate:"required"`
	Branch string `yaml:"branch" default:"main"`
}

// HubConfig is hub configuration with source registries.
type HubConfig struct {
	Sources []HubSource `yaml:"sources"`
}

// AgencyConfig is the schema for config.yaml.
type AgencyConfig struct {
	Hub HubConfig `yaml:"hub"`
}

// HubInstalledEntry tracks an installed hub component.
type HubInstalledEntry struct {
	Component   string `yaml:"component" validate:"required"`
	Kind        string `yaml:"kind" validate:"required"`
	Source      string `yaml:"source" validate:"required"`
	CommitSHA   string `yaml:"commit_sha" validate:"required"`
	InstalledAt string `yaml:"installed_at" validate:"required"`
}
